server/src/main/java/org/opensearch/action/get/TransportGetAction.java

```diff
@@ -88,16 +92,32 @@ protected boolean resolveIndex(GetRequest request) {
         return true;
     }
 
+    static boolean isSegmentReplicationEnabled(ClusterState state, String indexName) {
+        return Optional.ofNullable(state.getMetadata().index(indexName))
+            .map(
+                indexMetadata -> ReplicationType.parseString(indexMetadata.getSettings().get(IndexMetadata.SETTING_REPLICATION_TYPE))
+                    .equals(ReplicationType.SEGMENT)
+            )
+            .orElse(false);
+    }
+
+    /**
+     * Returns true if GET request should be routed to primary shards, else false.
+     */
+    protected static boolean isPrimaryBasedRouting(ClusterState state, boolean realtime, String preference, String indexName) {
+        return isSegmentReplicationEnabled(state, indexName) && realtime && preference == null;
+    }
+
     @Override
     protected ShardIterator shards(ClusterState state, InternalRequest request) {
+        String preference = request.request().preference();
```

Comment:
Nitpick, but assigning variables exactly once is often more readable and less error prone, e.g.:

```
final String preference;
if (isPrimary...) {
    preference = PRIMARY;
} else {
    preference = request.request().preference();
}
```
```